torch/csrc/inductor/aot_inductor_tensor.cpp

```diff
@@ -0,0 +1,185 @@
+        {kAotInductorCPU, c10::kCPU},
+        {kAotInductorCUDA, c10::kCUDA},
+};
+
+std::unordered_map<c10::DeviceType, AotInductorDeviceType>
+    C10DeviceType_to_AotInductorDeviceType = {
+        {c10::kCPU, kAotInductorCPU},
+        {c10::kCUDA, kAotInductorCUDA},
+};
+
+AotInductorScalarType convert_to_aot_inductor_scalar_type(
+    c10::ScalarType type) {
+  TORCH_CHECK(
+      C10ScalarType_to_AotInductorScalarType.count(type) != 0,
+      "ScalarType not supported by AotInductor");
+  return C10ScalarType_to_AotInductorScalarType.at(type);
+}
+
+c10::ScalarType convert_to_c10_scalar_type(AotInductorScalarType type) {
+  TORCH_CHECK(
+      AotInductorScalarType_to_C10ScalarType.count(type) != 0,
+      "ScalarType not supported by AotInductor");
+  return AotInductorScalarType_to_C10ScalarType.at(type);
+}
+
+AotInductorDevice convert_to_aot_inductor_device(c10::Device device) {
+  TORCH_CHECK(
+      C10DeviceType_to_AotInductorDeviceType.count(device.type()) != 0,
+      "DeviceType not supported by AotInductor");
+  AotInductorDevice result{
+      C10DeviceType_to_AotInductorDeviceType.at(device.type()), device.index()};
+  return result;
+}
+
+c10::Device convert_to_c10_device(AotInductorDevice device) {
+  TORCH_CHECK(
+      AotInductorDeviceType_to_C10DeviceType.count(device.device_type) != 0,
+      "DeviceType not supported by AotInductor");
+  c10::Device result{
+      AotInductorDeviceType_to_C10DeviceType.at(device.device_type),
+      device.device_id};
+  return result;
+}
+
+std::vector<std::vector<int64_t>> sizes_cache;
+std::vector<std::vector<int64_t>> strides_cache;
+
+const int64_t* register_sizes(c10::IntArrayRef sizes) {
+  sizes_cache.emplace_back(sizes.vec());
+  return sizes_cache[sizes_cache.size() - 1].data();
+}
+
+const int64_t* register_strides(c10::IntArrayRef strides) {
+  strides_cache.emplace_back(strides.vec());
+  return strides_cache[strides_cache.size() - 1].data();
+}
+
+} // namespace
+
+AotInductorTensor convert_to_aot_inductor_tensor(void* aten_tensor) {
+  at::Tensor* t = static_cast<at::Tensor*>(aten_tensor);
+
+  AotInductorTensor result = {
+      t->data_ptr(),
+      convert_to_aot_inductor_device(t->device()),
+      convert_to_aot_inductor_scalar_type(t->scalar_type()),
+      t->dim(),
+      register_sizes(t->sizes()),
+      register_strides(t->strides())};
+  return result;
+}
+
+void convert_to_aten_tensor(
+    AotInductorTensor inductor_tensor,
+    void* aten_tensor) {
+  c10::TensorOptions options =
+      c10::TensorOptions()
+          .device(convert_to_c10_device(inductor_tensor.device))
+          .dtype(convert_to_c10_scalar_type(inductor_tensor.type));
+
+  at::Tensor* t = static_cast<at::Tensor*>(aten_tensor);
+  (*t) = at::from_blob(
+      inductor_tensor.data_ptr,
+      c10::IntArrayRef(inductor_tensor.sizes, inductor_tensor.ndim),
+      c10::IntArrayRef(inductor_tensor.strides, inductor_tensor.ndim),
+      options);
```

Comment:
What is the memory management story here?  Who owns the memory?  How do we ensure it doesn't get freed before this tensor falls out of scope?

My understanding is `from_blob` doesn't have a refcount on the storage.

Reply:
The semantics here is that at::Tensor is a non-owning references to the AotInductorTensor, which sounds pretty difficult to use correctly. You probably need to pass around a function pointer that knows how to do deallocation and setup some move semantics, or arrange that the model so never returns any allocations (the runtime is expected to do all the allocations, and aot inductor only borrows.)

Reply:
One solution is to combine this with memory planning (https://github.com/pytorch/pytorch/pull/102625), so basically memory planner is our runtime wrt. memory.

Reply:
That would almost work, but there are fallback kernels like at::convolution() that don't have `out=` variants, so they return a newly-allocated tensors.

I think you need something like a `std::shared_ptr<AotTensorStorage>` to keep the memory allocated and reference count when it can be deleted.  Either that or you could manually increment/decrement the reference count on the at::TensorStorage.